### ISSUE TITLE
EPTHOOK causes process instruction execution exception。

### ISSUE THREAD

**Issue by Mack5443** [HyperDbg/HyperDbg#468]
```

//test.asm
.CODE
myFloatOperation PROC
    xorps xmm0,xmm0
    movsd xmm0, qword ptr [rdx]
    movaps xmm2, xmm0  ；put it here
    movsd qword ptr [rcx],xmm2
    ret
myFloatOperation ENDP

END
```
```


//main.cpp
EXTERN_C void myFloatOperation(double * a,double * b);
int main() {
    printf("%x\n", GetCurrentProcessId());
    while (true)
    {
        double a = 90.0;
        double b = 50.0;
        double* ap = &a;
        double* bp = &b;
        myFloatOperation(ap, bp);
        printf("a:%.2f  b:%.2f\n", a, b);
        Sleep(1000);
    }
   
}
```
**If I place the epthook on the instructions moveps xmm2 and xmm0, it will result in the instruction not being executed and the rip register not being modified in the script.**

```
HyperDbg> !epthook 7FF7CAA41007 pid 552c script {
>  printf("111\n");
>  }
```

The result of the execution is as follows:
```
//Effective
a:0.00  b:50.00
a:0.00  b:50.00
a:0.00  b:50.00
a:0.00  b:50.00
a:0.00  b:50.00
//Not effective
a:50.00  b:50.00
a:50.00  b:50.00
a:50.00  b:50.00
a:50.00  b:50.00
a:50.00  b:50.00
a:50.00  b:50.00
a:50.00  b:50.00
```

**Comment by Mack5443** [comment]
![Uploading Snipaste_2024-08-31_01-16-10.png…]()

**Comment by SinaKarvandi** [comment]
Hi,

Thanks for creating this issue. I created your exact same scenario but I couldn't reproduce your scenario. It's works flawlessly for me: 

![image](https://github.com/user-attachments/assets/45230454-90a0-4372-8aad-23f439901bf2)

**Comment by SinaKarvandi** [comment]
This is the code. Are you sure you're using the latest version of HyperDbg? Can you test it again using the 'dev' branch?

![image](https://github.com/user-attachments/assets/e7b88036-0cb0-4978-931d-02eec4259acf)

**Comment by SinaKarvandi** [comment]
Maybe your compiler is different than me? I use the latest version of MSVC.

**Comment by Mack5443** [comment]
Well, I'm using version 0.8.4 for testing. Maybe it works fine in the latest version.

**Comment by Mack5443** [comment]
I don't see that the operation of floating-point registers is not supported in the script commands, so I wonder if it is limited by some unavoidable constraint, or if it is not currently implemented

**Comment by SinaKarvandi** [comment]
> Well, I'm using version 0.8.4 for testing. Maybe it works fine in the latest version.

Ah, Got it. Yes, I can reproduce the error now on the latest version. I'll try to investigate and find the error. I'll update you here once it's fixed.

**Comment by SinaKarvandi** [comment]
> I don't see that the operation of floating-point registers is not supported in the script commands, so I wonder if it is limited by some unavoidable constraint, or if it is not currently implemented

Yes, there are actually technical constraints. By default, kernel code does not support floating point (even though it's possible in the processor), but since most of drivers don't need it, it's not supported by default. Also, to optimize the context switching, it's not even enabled by default for the user-mode applications. For the exact same reason, we won't save XMM or floating point registers in the VM-exit handler. Adding support to the floating point (even though it's possible) but it negatively impacts the overall HyperDbg performance in many cases.

**Comment by SinaKarvandi** [comment]
I investigated the issue and identified the root cause of the bug: the MSVC compiler uses XMM registers to optimize memory zeroing in the RtlZeroMemory function. It starts by zeroing the XMM0 registers, but since we avoid context switching for XMM registers (to reduce VM-exit overhead), this creates a problem.

A quick fix is to uncomment the relevant lines in `AsmVmexitHandler.asm`, but a more comprehensive solution requires further investigation.

https://github.com/HyperDbg/HyperDbg/blob/ceca0ae762b09ec9a8d55c94fcd62332e3657e04/hyperdbg/hyperhv/code/assembly/AsmVmexitHandler.asm#L20

![image](https://github.com/user-attachments/assets/d30140f9-ebd9-4a5f-95c7-39c495fd5f4f)

**Comment by Mack5443** [comment]
Thanks for investigating！



---Original---
From: "Sina ***@***.***&gt;
Date: Sun, Sep 8, 2024 20:34 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [HyperDbg/HyperDbg] EPTHOOK causes process instruction execution exception。 (Issue #468)




 
I investigated the issue and identified the root cause of the bug: the MSVC compiler uses XMM registers to optimize memory zeroing in the RtlZeroMemory function. It starts by zeroing the XMM0 registers, but since we avoid context switching for XMM registers (to reduce VM-exit overhead), this creates a problem.
 
A quick fix is to uncomment the relevant lines in AsmVmexitHandler.asm, but a more comprehensive solution requires further investigation.
 
https://github.com/HyperDbg/HyperDbg/blob/ceca0ae762b09ec9a8d55c94fcd62332e3657e04/hyperdbg/hyperhv/code/assembly/AsmVmexitHandler.asm#L20
 
image.png (view on web)

**Comment by SinaKarvandi** [comment]
The issue with this patch is that while it resolves the current problem, it significantly degrades the performance of nearly all HyperDbg functionalities. This is because the XMM registers would need to be saved and restored during every VM-exit and VM-entry, even though it's unnecessary in most cases. As a result, I'm unsure if applying this patch is the right solution. Ideally, we could prevent the MSVC compiler from using XMM registers, but it appears that this option isn't available to MSVC users.

**Comment by lauralex** [comment]
Actually there is an option to disable enhanced instructions in MSVC

**Comment by SinaKarvandi** [comment]
Do you mean this one? 

https://learn.microsoft.com/en-us/cpp/build/reference/arch-x64?view=msvc-170

It seems that by default it's enabled and we couldn't disable it. Am I missing something? Can you give an example to me?

**Comment by SinaKarvandi** [comment]
This problem is now fixed and will be available starting from the next version (v0.15). For now, you can use the 'dev' branch with the required modifications. (https://github.com/HyperDbg/HyperDbg/pull/541)
